flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/OperatorCoordinatorHolder.java

```diff
@@ -308,6 +335,26 @@ private void checkpointCoordinatorInternal(
         }
     }
 
+    private void closeGateways(final long checkpointId, final CompletableFuture<byte[]> result) {
+        Set<CloseableSubtaskGateway> closedGateways = new HashSet<>();
+        for (CloseableSubtaskGateway gateway : subtaskGatewayMap.values()) {
+            if (!gateway.tryCloseGateway(checkpointId)) {
+                closedGateways.forEach(CloseableSubtaskGateway::openGatewayAndUnmarkCheckpoint);
```

Comment:
Is there any case where `closedGateways` is not empty at this point? If no, could it be better to throw IllegalStateException if `closedGateways()` is not empty?

Reply:
I agree. I'll make it throw exception here.